gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.metrics.event.lineage;
+
+import java.util.Collection;
+import java.util.Map;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+
+import org.apache.gobblin.configuration.State;
+import org.apache.gobblin.dataset.DatasetDescriptor;
+import org.apache.gobblin.metrics.event.GobblinEventBuilder;
+
+
+/**
+ * The lineage coordinator in a Gobblin job with single source and multiple destinations
+ *
+ * <p>
+ *   In Gobblin, a work unit processes records from only one dataset. It writes output to one or more destinations,
+ *   depending on the number of branches configured in the job. One destination means an output as another dataset.
+ * </p>
+ *
+ * <p>
+ *   Lineage info is jointly collected from the source, represented by {@link org.apache.gobblin.source.Source} or
+ *   {@link org.apache.gobblin.source.extractor.Extractor}, and destination,
+ *   represented by {@link org.apache.gobblin.writer.DataWriter} or {@link org.apache.gobblin.publisher.DataPublisher}
+ * </p>
+ *
+ * <p>
+ *   The general flow is:
+ *   <ol>
+ *     <li> source sets its {@link DatasetDescriptor} to each work unit </li>
+ *     <li> destination puts its {@link DatasetDescriptor} to the work unit </li>
+ *     <li> load and send all lineage events from all states </li>
+ *     <li> purge lineage info from all states </li>
+ *   </ol>
+ * </p>
+ */
+public final class LineageInfo {
+  static final String BRANCH = "branch";
```

Comment:
Do you think these names like "branch or names" should have a namespace or scope? Otherwise user can overwrite these attribute in gobblin constructs, without noticing these attributes are important for sending lineage event. We need to document this behavior somewhere. Or we should have a scoped name.

Reply:
Not really. This is really a constant. It's not a name. And it's final, won't be changed.

Reply:
Actually. for job configurations, we do want to provide the flexibility where namespace can vary but keys are the same.

Reply:
What would happen if some components set "name=XYZ" in gobblin State object for other purpose instead of lineage event? Will it break the lineage event logic?